x/stakeibc/keeper/unbonding.go

```diff
@@ -309,6 +309,61 @@ func (k Keeper) GetUnbondingICAMessages(
 	return msgs, unbondings, nil
 }
 
+// Submit undelegate ICA messages in small batches to reduce the gas size per tx
+func (k Keeper) BatchSubmitUndelegateICAMessages(
+	ctx sdk.Context,
+	hostZone types.HostZone,
+	epochUnbondingRecordIds []uint64,
+	msgs []proto.Message,
+	unbondings []*types.SplitUndelegation,
+	batchSize int,
+) (numTxsSubmitted uint64, err error) {
+	// Iterate the full list of messages and submit in batches
+	for start := 0; start < len(msgs); start += batchSize {
+		end := start + batchSize
+		if end > len(msgs) {
+			end = len(msgs)
+		}
+
+		msgsBatch := msgs[start:end]
+		unbondingsBatch := unbondings[start:end]
+
+		// Store the callback data
+		undelegateCallback := types.UndelegateCallback{
+			HostZoneId:              hostZone.ChainId,
+			SplitUndelegations:      unbondingsBatch,
+			EpochUnbondingRecordIds: epochUnbondingRecordIds,
+		}
+		callbackArgsBz, err := proto.Marshal(&undelegateCallback)
+		if err != nil {
+			return numTxsSubmitted, errorsmod.Wrap(err, "unable to marshal undelegate callback args")
+		}
+
+		// Submit the undelegation ICA
+		if _, err := k.SubmitTxsDayEpoch(
```

Comment:
Now that we are batching these together a particular undelegation ICA can contain msgs from different records, all of these will succeed or fail together because a level or two up the whole call to UnbondFromHostZone is inside of a utils.ApplyFuncIfNoError -- so if there is any hostZoneRecord which has a persistent problem unbonding, this batching will now mean all will be blocked correct?  There is no way that part of a hostZoneRecord can succeed in unbonding while other messages from the same hostZoneRecord fail, all will succeed or fail together?

Reply:
so the ApplyFuncIfNoError is wrapped around the ICA _submission_ - so there's atomicity in that either all ICAs will fire or none of them will

But once they all fire, we no longer have atomicity with the responses. That's where certain batches can succeed and others can fail (hence why we have the retry logic and why those tokens are pooled with the new unbondings next epoch).

Reply:
This is the super key info.  As you had previously clarified on slack:

- Every epoch we send 20 batches of 5 undelegate messages, for a total of 100 messages
- This means each ICA tx contains 5 messages
- Submission for all batches/txs/messages is atomic
- Ack for each message within a tx/batch is atomic
- Ack for different txs/batches are not atomic

So importantly individual batches can fail or succeed separate from each other even though the whole batch itself will have the same fate.  I'm going to do another pass with this in mind since this is so crucial to the whole design.